src/cursor/cur_hs.c

```diff
@@ -511,78 +509,33 @@ __curhs_search_near(WT_CURSOR *cursor, int *exactp)
     WT_ERR(__wt_hs_cursor_search_near(session, file_cursor, &exact));
 
     /*
-     * There are some key fields missing so we are searching a range of keys. Place the cursor at
-     * the start of the range.
+     * If we have some missing fields in the key, try to walk forward first to increase the chance
+     * of placing the cursor at the key that is larger or equal to the search key. If nothing is
+     * found, walk backward.
      */
     if (!F_ISSET(hs_cursor, WT_HS_CUR_COUNTER_SET)) {
-        /*
-         * If we raced with a history store insert, we may be two or more records away from our
-         * target. Keep iterating forwards until we are on or past our target key.
-         *
-         * We can't use the cursor positioning helper that we use for regular reads since that will
-         * place us at the end of a particular key/timestamp range whereas we want to be placed at
-         * the beginning.
-         */
-        if (exact < 0) {
-            while ((ret = __wt_hs_cursor_next(session, file_cursor)) == 0) {
-                WT_ERR(__wt_compare(session, NULL, &file_cursor->key, srch_key, &cmp));
-                if (cmp >= 0)
-                    break;
-            }
-            /*
-             * No entries greater than or equal to the key we searched for. Reset cursor if we get
-             * WT_NOTFOUND.
-             */
-            WT_ERR(ret);
-
-            *exactp = cmp;
+        WT_ERR_NOTFOUND_OK(__curhs_next_visible(session, hs_cursor), true);
+        if (ret == WT_NOTFOUND) {
+            WT_ERR(__wt_hs_cursor_prev(session, file_cursor));
```

Comment:
If there is a WT_NOTFOUND for next() call, wouldn't it reset the cursor?

Reply:
It uses the underlying file cursor which is not reseted.

Reply:
That's what I wanted to confirm - underlying cursor doesn't reset when the `next()` or `prev()` call returns `WT_NOTFOUND`.

Reply:
In the underlying cursor, if it returns not found, we are at the end of the tree and call next() will place us at the start of the tree and call prev() will place us on the last key of the tree(). In our case, apart from walking to the end of the tree, we also return not found in hs cursor if we cross the btree and key space. However, the underlying cursor in this case is still placed but on a key from a different btree or key. So calling next() and prev() will be fine.

Reply:
Thanks for explaining this Chenhao!

Reply:
Your comment reminds me that there is one issue. We can insert more contents to the hs concurrently and calling one next or prev doesn't guarantee we can go back the previous key space. I'll rectify that.